src/org/mozilla/javascript/NativeReflect.java

```diff
@@ -0,0 +1,419 @@
+                    "3",
+                    Integer.toString(args.length));
+        }
+
+        Scriptable callable = ScriptableObject.ensureScriptable(args[0]);
+
+        if (args[1] instanceof Scriptable) {
+            thisObj = (Scriptable) args[1];
+        } else if (ScriptRuntime.isPrimitive(args[1])) {
+            thisObj = cx.newObject(scope, "Object", new Object[] {args[1]});
+        }
+
+        if (ScriptRuntime.isSymbol(args[2])) {
+            throw ScriptRuntime.typeErrorById("msg.arg.not.object", ScriptRuntime.typeof(args[2]));
+        }
+        ScriptableObject argumentsList = ScriptableObject.ensureScriptableObject(args[2]);
+
+        return ScriptRuntime.applyOrCall(
+                true, cx, scope, callable, new Object[] {thisObj, argumentsList});
+    }
+
+    /** see https://262.ecma-international.org/12.0/#sec-reflect.construct */
+    private static Scriptable construct(
+            Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        /*
+         * 1. If IsConstructor(target) is false, throw a TypeError exception.
+         * 2. If newTarget is not present, set newTarget to target.
+         * 3. Else if IsConstructor(newTarget) is false, throw a TypeError exception.
+         * 4. Let args be ? CreateListFromArrayLike(argumentsList).
+         * 5. Return ? Construct(target, args, newTarget).
+         */
+        if (args.length < 1) {
+            throw ScriptRuntime.typeErrorById(
+                    "msg.method.missing.parameter",
+                    "Reflect.construct",
+                    "3",
+                    Integer.toString(args.length));
+        }
+
+        if (!(args[0] instanceof Constructable)) {
+            throw ScriptRuntime.typeErrorById("msg.not.ctor", ScriptRuntime.typeof(args[0]));
+        }
+
+        Constructable ctor = (Constructable) args[0];
+        if (args.length < 2) {
+            return ctor.construct(cx, scope, ScriptRuntime.emptyArgs);
+        }
+
+        if (args.length > 2 && !(args[2] instanceof Constructable)) {
+            throw ScriptRuntime.typeErrorById("msg.not.ctor", ScriptRuntime.typeof(args[2]));
+        }
+
+        Object[] callArgs = ScriptRuntime.getApplyArguments(cx, args[1]);
+
+        Object newTargetPrototype = null;
+        if (args.length > 2) {
+            Scriptable newTarget = ScriptableObject.ensureScriptable(args[2]);
+
+            if (newTarget instanceof BaseFunction) {
+                newTargetPrototype = ((BaseFunction) newTarget).getPrototypeProperty();
+            } else {
+                newTargetPrototype = newTarget.get("prototype", newTarget);
+            }
+
+            if (!(newTargetPrototype instanceof Scriptable)
+                    || ScriptRuntime.isSymbol(newTargetPrototype)
+                    || Undefined.isUndefined(newTargetPrototype)) {
+                newTargetPrototype = null;
+            }
+        }
+
+        // hack to set the right prototype before calling the ctor
```

Comment:
Maybe elaborate a bit on why this is considered a hack/why this hack is needed?

Reply:
done